### DISCUSSION TITLE
Customizing the disclosure to add slots for actions

### DISCUSSION THREAD

**Discussion by msaunders-twilio** [twilio-labs/paste#3473] [Office hours] [answered]
### Background

We are building AB experimentation into Segment's Broadcasts.  Part of this feature is showing the user all the different templates they are using and some meta data associated with each template during the creation of their Broadcast.  Given there is a lot of information on the page, we have chosen to use a Disclosure component to contain and show a template so that it can be minimized.

One of the things our designs include are action to duplicate or delete a template from the experiment.  We think those actions should live in the heading of the Disclosure.  However, when trying to buttons to the heading, clicking the button also triggers the opening and closing of the disclosure (see recording below).

https://github.com/twilio-labs/paste/assets/109765412/eb338348-c38a-47d0-b271-bed7af665c66

### Feedback

We would like to enable the user to click an action without triggering the opening and closing of the Disclosure.  Is what we are trying to accomplish feasible with the Disclosure component?  

If not, how could we update the Disclosure component to enable this use case?  We would be happy to guest engineer to make this happen if a change is required.

### Stage of design

High-Fidelity.  A screenshot of our design for this section is included below.

<img width="376" alt="Screenshot 2023-09-06 at 9 39 30 AM" src="https://github.com/twilio-labs/paste/assets/109765412/a885a6f3-c0f1-4699-bae1-9d87f23a5cf8">

### Relevant links

Figma:
https://www.figma.com/file/kp9Jout2yMO6IKE6MbfpVG/%5BDesign%5D-Experimentation-%E2%80%93-Broadcasts-and-messages?type=design&node-id=723-45282&mode=design&t=AavIEIqs3E2tyVQM-0

### Are you looking for visual design support?

- [ ] Yes

**Comment by ghost** [comment]
In addition to general feasibility, would appreciate additional feedback and input on the **accessibility** of this approach. Is there a challenge around having nested selectable/focusable elements within each other? If so, how might we best include actions with each disclosure object?

**Comment by SiTaggart** [comment, marked as answer]
Hi there

As per the [Disclosure documentation](https://paste.twilio.design/components/disclosure#accessibility), nested focusable elements are not accessible and should not be implemented.

> You **must not** nest focusable or actionable elements inside the `DisclosureHeading`. Nested actionable elements are difficult to discover for users of assistive technology as their existence is often not announced.

You can do something pretty similar though which would be perfectly accessible, using the [Disclosure Primitive](https://paste.twilio.design/primitives/disclosure-primitive#about-the-disclosure-primitive).

We've planned to build something called a Summary Detail component for a while, but it's never quite made the prioritization cut. I've knocked up a [quick Codesandbox](https://codesandbox.io/s/summary-detail-example-n3xsh3?file=/src/index.tsx) to demo what it might look like. 

You might take some inspiration from that and build something yourselves, that would be perfectly accessible and so the same job as your design intends.

<img width="996" alt="image" src="https://github.com/twilio-labs/paste/assets/368249/b25f97cb-c6e3-4342-9aa9-b3977ba6b125">

Essentially, the trigger is separated from the Heading as a sibling, allowing you to have as many actions in line as you like, but still be collapsible.

**Reply by ghost** [reply]
Perfect, thanks @SiTaggart! We'll follow your suggestion here.
